File: tests/contract/memref_3way.c
/**
 * memref_3way.c — 3-way MemRef layout contract test (C side)
 *
 * Verifies that sizeof(SFATensorRaw1-4) and field offsets in
 * include/sfa.h match the expected MLIR memref descriptor layout:
 *   struct<(ptr, ptr, i64, array<RANK x i64>, array<RANK x i64>)>
 *
 * Output format (one value per line):
 *   C:40
 *   C:56
 *   C:72
 *   C:88
 *   C:allocated=0
 *   C:aligned=8
 *   C:offset=16
 *
 * Build: cc -I../../include -o memref_3way_c memref_3way.c
 */

#include <stddef.h>
#include <stdio.h>
#include "sfa.h"

int main(void) {
    /* sizeof assertions (must match the header comment + Rust + Python) */
    printf("C:%zu\n", sizeof(SFATensorRaw1));
    printf("C:%zu\n", sizeof(SFATensorRaw2));
    printf("C:%zu\n", sizeof(SFATensorRaw3));
    printf("C:%zu\n", sizeof(SFATensorRaw4));

    /* Field offset assertions for the 24-byte memref header.
     * SFATensorRaw2 is used as the canonical rank since the header
     * layout is identical across all ranks. */
    printf("C:allocated=%zu\n", offsetof(SFATensorRaw2, allocated));
    printf("C:aligned=%zu\n",   offsetof(SFATensorRaw2, aligned));
    printf("C:offset=%zu\n",    offsetof(SFATensorRaw2, offset));

    return 0;
}

File: tests/contract/memref_3way.py
"""
memref_3way.py — 3-way MemRef layout contract test (Python side)

Verifies that the per-rank SFATensorRaw struct sizes and field offsets
match the C ABI from include/sfa.h using struct.Struct.

Output format (one value per line, same keys as C and Rust):
  Python:40
  Python:56
  Python:72
  Python:88
  Python:allocated=0
  Python:aligned=8
  Python:offset=16
"""

import struct
import sys


def main() -> int:
    ptr_size = struct.Struct("P").size  # void* — 8 bytes on 64-bit
    i64_size = struct.Struct("q").size  # int64_t — 8 bytes
    header = ptr_size + ptr_size + i64_size  # allocated + aligned + offset

    # Compute sizes: header + rank * 8 (sizes array) + rank * 8 (strides array)
    sizes = {}
    for rank in range(1, 5):
        total = header + 2 * rank * i64_size
        sizes[rank] = total
        print(f"Python:{total}")

    # Field offsets for the 24-byte memref header (SFATensorRaw2).
    # The layout is struct<(void*, void*, int64_t, int64_t[2], int64_t[2])>
    assert struct.calcsize("PPq2q2q") == 56, f"SFATensorRaw2 expected 56, got {struct.calcsize('PPq2q2q')}"
    print("Python:allocated=0")
    print(f"Python:aligned={ptr_size}")
    print(f"Python:offset={ptr_size + ptr_size}")

    return 0


if __name__ == "__main__":
    sys.exit(main())
